core/src/main/scala/astraea/spark/rasterframes/expressions/tilestats/ForAll.scala

```diff
@@ -0,0 +1,55 @@
+package astraea.spark.rasterframes.expressions.tilestats
+
+import astraea.spark.rasterframes.expressions.UnaryRasterOp
+import astraea.spark.rasterframes.model.TileContext
+import geotrellis.raster.{Tile, isData}
+import org.apache.spark.sql.{Column, TypedColumn}
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription}
+import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
+import org.apache.spark.sql.types._
+
+@ExpressionDescription(
+  usage = "_FUNC_(tile) - Returns true if all cells in the tile are true (non-zero and not nodata).",
+  arguments =
+    """
+    Arguments:
+       * tile - tile to check
+    """,
+  examples =
+    """
+    > SELECT _FUNC_(tile);
+       true
+    """
+)
+case class ForAll(child: Expression) extends UnaryRasterOp with CodegenFallback {
+  override def nodeName: String = "for_all"
+  override def dataType: DataType = BooleanType
+  override protected def eval(tile: Tile, ctx: Option[TileContext]): Any = ForAll.op(tile)
+
+}
+
+object ForAll{
+  import astraea.spark.rasterframes.encoders.StandardEncoders.PrimitiveEncoders.boolEnc
+
+  def apply(tile: Column): TypedColumn[Any, Boolean] = new Column(ForAll(tile.expr)).as[Boolean]
+
+  def op(tile: Tile): Boolean = {
+
+    def doubleValueIs(d: Double): Boolean =  isData(d) & d != 0.0
+    def intValueIs(i: Int): Boolean =  isData(i) & i != 0
+
+    var (c, r) = (0, 0)
+    while (r < tile.rows) {
+      while(c < tile.cols) {
+        if(tile.cellType.isFloatingPoint) { if(! doubleValueIs(tile.getDouble(c, r))) return false }
+        else { if(! intValueIs(tile.get(c, r))) return false }
+        c += 1
+      }
+      c = 0; r += 1
+    }
```

Comment:
Feels like there's some code duplication going on; in two areas, one subtle, the other more obvious. 

The subtle one is that there should be a single place in the system where we define what is interpreted as `true` and `false` in the context of a cell (just like GeoTrellis does with `isNoData`. I'd suggest putting functions in the `rasterframes` package object. Maybe something like one of these:

* `cellAsBoolean(v: {Int|Double})`
* `interpretedAsBoolean(v: {Int|Double})`
* `isTrueish(v: {Int|Double})` (or `isTruthy` ;-) )
* `isCellTrue(v: {Int|Double})`
* `isDataTrue(v: {Int|Double})`


Remember that in Scala you have function overloading, so you don't need to prefix the function with `int` or `double`.

The other duplication area is that `op` implementations seem very similar in structure, but I admit that the short circuit evaluation you're doing may make it hard to refactor.

Reply:
I have no real vision on how to reduce apparent duplication of the short circuited loops here. But point very much taken on the other.